drivers/gpio/gpio_stmpe1600.c

```diff
@@ -0,0 +1,297 @@
+	if (ret) {
+		return ret;
+	}
+
+	if ((flags & GPIO_INPUT) != 0U) {
+		reg_dir &= ~BIT(pin);
+	} else {
+		reg_dir |= BIT(pin);
+		if ((flags & GPIO_OUTPUT_INIT_HIGH) != 0U) {
+			reg_state |= BIT(pin);
+		} else if ((flags & GPIO_OUTPUT_INIT_LOW) != 0U) {
+			reg_state &= ~BIT(pin);
+		}
+	}
+
+	ret = write_reg16(dev, REG_GPDR_LSB, reg_dir);
+	if (ret) {
+		return ret;
+	}
+	ret = write_reg16(dev, REG_GPSR_LSB, reg_state);
+	return ret;
+}
+
+static int stmpe1600_init(const struct device *dev)
+{
+	const struct stmpe1600_config *const config = dev->config;
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	const struct device *i2c_master;
+	uint16_t chip_id = 0;
+	int ret = 0;
+
+	LOG_DBG("STMPE1600[0x%02X] init", config->i2c_slave_addr);
+
+	i2c_master = device_get_binding((char *)config->i2c_master_dev_name);
+	if (!i2c_master) {
+		return -EINVAL;
+	}
+	drvdata->i2c_master = i2c_master;
+
+	k_sem_init(&drvdata->lock, 1, 1);
+
+	k_sem_take(&drvdata->lock, K_FOREVER);
+	ret = read_reg16(dev, REG_CHIP_ID_LSB, &chip_id);
+	k_sem_give(&drvdata->lock);
+
+	if (ret) {
+		LOG_ERR("STMPE1600[0x%02X]: Unable to read Chip ID", config->i2c_slave_addr);
+		return ret;
+	}
+
+	if (chip_id != 0x1600) {
+		LOG_ERR("STMPE1600[0x%02X]: Invalid Chip ID", config->i2c_slave_addr);
+		return -EINVAL;
+	}
+
+	return ret;
+}
+
+static int stmpe1600_configure(const struct device *dev,
+			       gpio_pin_t pin, gpio_flags_t flags)
+{
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	const struct stmpe1600_config *const config = dev->config;
+	uint16_t i2c_addr = config->i2c_slave_addr;
+	int ret = 0;
+
+	/* No support for disconnected pin */
+	if ((flags & (GPIO_INPUT | GPIO_OUTPUT)) == GPIO_DISCONNECTED) {
+		return -ENOTSUP;
+	}
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drvdata->lock, K_FOREVER);
+
+	ret = set_pin_dir(dev, pin, flags);
+	if (ret) {
+		LOG_ERR("STMPE1600[0x%X]: error setting pin direction (%d)", i2c_addr, ret);
+	}
+
+	k_sem_give(&drvdata->lock);
+	return ret;
+}
+
+static int stmpe1600_port_get_raw(const struct device *dev, uint32_t *value)
+{
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const) dev->data;
+	int ret = 0;
+	uint16_t reg_value = 0;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drvdata->lock, K_FOREVER);
+
+	ret = read_reg16(dev, REG_GPMR_LSB, &reg_value);
+	if (ret == 0) {
+		*value = reg_value;
+	}
+
+	k_sem_give(&drvdata->lock);
+	return ret;
+}
+
+static int stmpe1600_port_set_masked_raw(const struct device *dev,
+					 uint32_t mask, uint32_t value)
+{
+	struct stmpe1600_drvdata *const drvdata = (struct stmpe1600_drvdata *const)dev->data;
+	uint16_t buf;
+	int ret = 0;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drvdata->lock, K_FOREVER);
+
+	if (mask != 0xffff) {
+		ret = read_reg16(dev, REG_GPSR_LSB, &buf);
```

Comment:
It would be better to cache the state of the register in the driver. Reading the value of the register over I2C every time is not efficient.